pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -94,62 +98,96 @@ func (d *Dynamic) patternsToDetectors(patterns GCOMPatterns) ([]angulardetector.
+// If etag is not empty, it will be sent as If-None-Match header.
+// If the response status code is 304, it returns errNotModified.
+func (d *Dynamic) fetch(ctx context.Context, etag string) (GCOMResponse, error) {
 	st := time.Now()
 
 	reqURL, err := url.JoinPath(d.baseURL, gcomAngularPatternsPath)
 	if err != nil {
-		return nil, fmt.Errorf("url joinpath: %w", err)
+		return GCOMResponse{}, fmt.Errorf("url joinpath: %w", err)
 	}
 
 	d.log.Debug("Fetching dynamic angular detection patterns", "url", reqURL)
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
 	if err != nil {
-		return nil, fmt.Errorf("new request with context: %w", err)
+		return GCOMResponse{}, fmt.Errorf("new request with context: %w", err)
+	}
+	if etag != "" {
+		req.Header.Add("If-None-Match", etag)
 	}
+
+	var r GCOMResponse
 	resp, err := d.httpClient.Do(req)
 	if err != nil {
-		return nil, fmt.Errorf("http do: %w", err)
+		return GCOMResponse{}, fmt.Errorf("http do: %w", err)
 	}
 	defer func() {
 		if closeErr := resp.Body.Close(); closeErr != nil {
 			d.log.Error("Response body close error", "error", err)
 		}
 	}()
+	if resp.StatusCode == http.StatusNotModified {
+		return GCOMResponse{}, errNotModified
+	}
 	if resp.StatusCode/100 != 2 {
-		return nil, fmt.Errorf("bad status code: %d", resp.StatusCode)
+		return GCOMResponse{}, fmt.Errorf("bad status code: %d", resp.StatusCode)
 	}
-	var out GCOMPatterns
-	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
-		return nil, fmt.Errorf("json decode: %w", err)
+	if err := json.NewDecoder(resp.Body).Decode(&r.Patterns); err != nil {
+		return GCOMResponse{}, fmt.Errorf("json decode: %w", err)
 	}
-	d.log.Debug("Fetched dynamic angular detection patterns", "patterns", len(out), "duration", time.Since(st))
-	return out, nil
+	r.ETag = resp.Header.Get("ETag")
+	d.log.Debug("Fetched dynamic angular detection patterns", "patterns", len(r.Patterns), "duration", time.Since(st))
+	return r, nil
 }
 
 // updateDetectors fetches the patterns from GCOM, converts them to detectors,
 // stores the patterns in the database and update the cached detectors.
-func (d *Dynamic) updateDetectors(ctx context.Context) error {
+func (d *Dynamic) updateDetectors(ctx context.Context, etag string) error {
 	// Fetch patterns from GCOM
 	d.mux.Lock()
 	defer d.mux.Unlock()
-	patterns, err := d.fetch(ctx)
-	if err != nil {
+
+	resp, err := d.fetch(ctx, etag)
+	switch {
+	case err == nil:
+		break
+	case errors.Is(err, errNotModified):
+		// If patterns did not change, update the last updated anyways,
+		// so we don't keep trying to fetch them, but update the in-memory
+		// detectors from the previously stored value.
+		d.log.Debug("Not modified, skipping update")
+		if err := d.store.SetLastUpdated(ctx); err != nil {
+			return fmt.Errorf("set last updated: %w", err)
+		}
+		// Update in-memory detectors, by reading current value in the kvstore
+		if err := d.setDetectorsFromCache(ctx); err != nil {
```

Comment:
just to double check, would it be possible that the server returns a 304 while the database is empty? (e.g. a new instance?)

Reply:
The patterns are hardcded in GCOM, so it should never happen. But even if we consider an empty list, it is treated just like any other value, so it should return a different ETag to identify an empty response and it should keep working